Projects/Server/Text/TextEncoding.cs

```diff
@@ -114,33 +115,55 @@ public static int GetByteLengthForEncoding(this Encoding encoding) =>
             _          => 1
         };
 
-    [MethodImpl(MethodImplOptions.AggressiveInlining)]
-    private static bool IsSafeChar(ushort c) => c is >= 0x20 and < 0xFFFE;
-
     public static string GetString(ReadOnlySpan<byte> span, Encoding encoding, bool safeString = false)
     {
-        string s = encoding.GetString(span);
-
         if (!safeString)
         {
-            return s;
+            return encoding.GetString(span);
         }
 
-        ReadOnlySpan<char> chars = s.AsSpan();
+        var charCount = encoding.GetMaxCharCount(span.Length);
 
-        using var sb = new ValueStringBuilder(stackalloc char[256]);
-        var hasDoneAnyReplacements = false;
+        char[] rentedChars = null;
+        Span<char> chars = charCount <= 256
+            ? stackalloc char[charCount]
+            : rentedChars = STArrayPool<char>.Shared.Rent(charCount);
 
-        for (int i = 0, last = 0; i < chars.Length; i++)
+        try
         {
-            if (!IsSafeChar(chars[i]) && i == chars.Length - 1)
+            var length = encoding.GetChars(span, chars);
+            chars = chars[..length];
```

Comment:
This still allocates with weird strings. I am not going to spend the time to guard against malicious input that could cause 2-5x perf hit or the 300+ byte allocation. It's unfortunate, but maybe someone knows of a way to check/sanitize a buffer for real characters and effectively bail out early if the string is corrupt/nonsense.